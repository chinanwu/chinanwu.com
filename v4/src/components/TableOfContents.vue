<template>
  <nav class="TableOfContents --uppercase" aria-labelledby="toc">
    <h2 class="headerLabel" id="toc">TABLE OF CONTENTS</h2>
    <ol class="TableOfContents__list" tabindex="0" @keydown="handleKeyDown">
      <li v-for="(content, index) in contents" :key="'tOC-' + index">
        <strong>
          <a :href="'#' + content.link" :title="content.title">{{
            content.label
          }}</a>
        </strong>
      </li>
    </ol>
  </nav>
</template>

<script>
export default {
  name: "TableOfContents",
  props: {
    contents: {
      type: Array,
      required: true,
      validator: values =>
        values.every(
          value =>
            value.label &&
            value.link &&
            (typeof value.label === "string" ||
              value.label instanceof String) &&
            (typeof value.link === "string" || value.link instanceof String)
        )
    }
  },
  methods: {
    handleKeyDown(event) {
      if (
        event?.key &&
        !event.shiftKey &&
        !event.ctrlKey &&
        !event.altKey &&
        !event.metaKey
      ) {
        // TODO: Look into implementing HOME and END? Perhaps
        if (/^[0-9]$/.test(event.key) && event.key <= this.contents.length) {
          event.preventDefault();
          document.location = `/#${this.contents[event.key - 1].link}`;
          document.getElementById(this.contents[event.key - 1].focusId).focus();
        }
      }
    }
  }
};
</script>

<style lang="less" scoped>
.TableOfContents {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.TableOfContents__list {
  font-size: 3.4rem;
}

.TableOfContents__list {
  padding: 0;
  margin: 0;
  line-height: 200%;
}

@media only screen and (max-width: 595px) {
  .TableOfContents__list {
    font-size: 3rem;
  }
}

@media only screen and (max-width: 320px) {
  .TableOfContents__list {
    font-size: 2.8rem;
  }
}
</style>
