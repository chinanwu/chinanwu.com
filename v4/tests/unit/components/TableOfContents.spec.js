import { shallowMount, mount } from "@vue/test-utils";
import TableOfContents from "@/components/TableOfContents.vue";

describe("TableOfContents", () => {
  describe("rendering", () => {
    it("renders", () => {
      const wrapper = shallowMount(TableOfContents, {
        props: {
          contents: [
            { label: "Test", link: "test", title: "Fake link", focusId: "test" }
          ]
        }
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe("behaviour", () => {
    xit("jumps to proper location", () => {
      document.createElement("div").setAttribute("id", "test");
      const wrapper = mount(TableOfContents, {
        props: {
          contents: [
            { label: "Test", link: "test", title: "Fake link", focusId: "test" }
          ]
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      wrapper.find("#tableOfContents").trigger("keydown", {
        key: 1
      });
    });
  });
});
