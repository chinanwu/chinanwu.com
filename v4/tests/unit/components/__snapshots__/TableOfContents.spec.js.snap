// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableOfContents rendering renders 1`] = `
<nav
  aria-labelledby="toc"
  class="TableOfContents --uppercase"
>
  <h2
    class="headerLabel"
    id="toc"
  >
    Table of Contents
  </h2>
  <ol
    class="TableOfContents__list"
    id="tableOfContents"
    tabindex="0"
  >
    
    <li>
      <strong>
        <a
          class="--uppercase"
          href="#test"
          title="Fake link"
        >
          Test
        </a>
      </strong>
    </li>
    
  </ol>
</nav>
`;
